src/python/pants/jvm/resources.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.core.util_rules.archive import ZipBinary
+from pants.core.util_rules.source_files import SourceFilesRequest
+from pants.core.util_rules.stripped_source_files import StrippedSourceFiles
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import SourcesField
+from pants.engine.unions import UnionRule
+from pants.jvm.compile import (
+    ClasspathEntry,
+    ClasspathEntryRequest,
+    CompileResult,
+    FallibleClasspathEntry,
+)
+from pants.jvm.target_types import JvmResourcesFieldSet, JvmResourcesGeneratorFieldSet
+
+
+class JvmResourcesRequest(ClasspathEntryRequest):
+    field_sets = (
+        JvmResourcesFieldSet,
+        JvmResourcesGeneratorFieldSet,
+    )
+
+
+_PANTS_RESOURCES_PARTIAL_JAR_FILENAME = "pants-resources.jar"
+
+
+@rule(desc="Fetch with coursier")
+async def assemble_resources_jar(
+    bash: BashBinary,
+    zip: ZipBinary,
+    request: JvmResourcesRequest,
+) -> FallibleClasspathEntry:
+
+    source_files = await Get(
+        StrippedSourceFiles,
+        SourceFilesRequest([tgt.get(SourcesField) for tgt in request.component.members]),
+    )
+
```

Comment:
Since `resources` targets have a `Dependencies` field, you'll need to consume `request.component.dependencies` as well, which looks like this for `scalac`: https://github.com/pantsbuild/pants/blob/7e9d70da89b6bb46c42171cfb2ccd42bb94d7a1b/src/python/pants/backend/scala/compile/scalac.py#L56-L75

But this makes the optimization I mentioned below (build a single `jar` for a `resources` target) a little more complicated: the `resources` target will depend directly on a `resource` target per file: so the optimization would probably want to compute a `ClasspathEntryRequest` per dependency, and then _skip_ (completely drop on the floor) any for `resource` targets.

Reply:
It looks like that dependency fetching code is repeated in quite a lot of places at this point. Worth a refactor?

Reply:
@stuhood It's not clear to me that dropping `resource` targets is the right thing here: say we have 

```
resources(name="a", sources=["a.txt", "b.txt"], dependencies=[":c"])
resource(name="c", sources=["c.txt"])
```

then dropping `c` is probably not the right thing to do here. 

Is the correct solution to only skip over `resource` targets that were specifically generated by the current Generator target?

Reply:
> Is the correct solution to only skip over `resource` targets that were specifically generated by the current Generator target?

Yea, that's right. Thanks!

Reply:
@stuhood OK, so the complication here is that all the `dependencies` are `CoarsenedTarget`s -- is it a reasonable approach to reconstruct each `CoarsenedTarget` with the problem members filtered out, or is there a saner way?

Reply:
> @stuhood OK, so the complication here is that all the `dependencies` are `CoarsenedTarget`s -- is it a reasonable approach to reconstruct each `CoarsenedTarget` with the problem members filtered out, or is there a saner way?

It would probably be totally fine to skip the optimization if any `CoarsenedTarget` is not entirely `resource`s. It should be exceedingly rare that resources are involved in a cycle.

Reply:
@stuhood I think the solution here is correct now.